django-stubs/contrib/auth/backends.pyi

```diff
@@ -1,19 +1,18 @@
 from typing import Any, TypeVar
 
-from django.contrib.auth.base_user import AbstractBaseUser
+from django.contrib.auth import get_user_model
 from django.contrib.auth.models import AnonymousUser, Permission
 from django.db.models import QuerySet
 from django.db.models.base import Model
 from django.http.request import HttpRequest
 from typing_extensions import TypeAlias
 
-_AnyUser: TypeAlias = AbstractBaseUser | AnonymousUser
-
-UserModel: Any
+_UserModel = get_user_model()
```

Comment:
Can we declare `_UserModel` in 1 file and import it in the others? Or does `_UserModel` exist runtime?

Then we should probably keep it in `contrib.auth` next to definition of `get_user_model`

Reply:
Like this:

```python
from django.contrib.auth import _UserModel as _UserModel
```

Reply:
`_UserModel` does not exist at runtime, but in some modules `UserModel` does exist, for example in [django/contrib/auth/backens.py](https://github.com/django/django/blob/main/django/contrib/auth/backends.py#L5). The subtest complains about these though, saying the typing doesn't match runtime for some attributes

Reply:
Yes, I also think we should import from one place it and not define for each module.

Reply:
I gave this approach a go now, but for some reason it ends up causing an endless defer look, which crashes mypy. Can we leave it like this for now and maybe revisit? It does kinda seem like a bug in mypy, cause we never get a call with `final_iteration=True` before it gives up, but I'm not entirely sure

Reply:
I did a step by step removal on your branch and realised that it's possible to replace with an import in all files except `django.http.request`.

Which is imported by e.g. `django.contrib.auth` and `django.contrib.auth.backends`. So I'm guessing that makes a bidirectional dependency?

`django.contrib.auth` imports `django.http.request`, while `django.http.request` imports `django.contrib.auth`..

Reply:
I haven't run the whole suite just yet, but I tried this diff:

```diff
         # for `get_user_model()`
-        if self.django_context.settings and any(
-            i.id == "django.contrib.auth" and any(name == "get_user_model" for name, _ in i.names)
-            for i in file.imports
-            if isinstance(i, ImportFrom)
-        ):
+        if file.fullname == "django.contrib.auth" and self.django_context.settings:
             auth_user_model_name = self.django_context.settings.AUTH_USER_MODEL
             ...
```

Together with these changes:

```diff
-from django.contrib.auth import get_user_model
+from django.contrib.auth import _UserModel
...

-_UserModel = get_user_model()
```

And an additional test case that depends on `django.contrib.auth` to declare a custom user, while also testing that the custom user is resolved via `get_user_model` by itself.

```yaml
-   case: test_user_model_resolved_via_get_user_model
    main: |
        from typing import Union
        from django.contrib.auth import get_user_model
        from django.contrib.auth.models import AnonymousUser
        from django.http import HttpRequest, HttpResponse
        from django.contrib.auth.decorators import user_passes_test
        User = get_user_model()
        reveal_type(User)
        def check(user: Union[User, AnonymousUser]) -> bool: return True
        @user_passes_test(check)
        def view(request: HttpRequest) -> HttpResponse:
            reveal_type(request.user)
            return HttpResponse()
    out: |
        main:7: note: Revealed type is "def (*args: Any, **kwargs: Any) -> myapp.models.MyUser"
        main:11: note: Revealed type is "Union[myapp.models.MyUser, django.contrib.auth.models.AnonymousUser]"
    custom_settings: |
        INSTALLED_APPS = ("django.contrib.contenttypes", "django.contrib.auth", "myapp")
        AUTH_USER_MODEL = "myapp.MyUser"
    files:
        - path: myapp/__init__.py
        - path: myapp/models.py
          content: |
              from django.contrib.auth.models import AbstractUser, PermissionsMixin

              class MyUser(AbstractUser, PermissionsMixin):
                  ...
```

I get both your test and the one above passing with these changes

Reply:
I've seen things crash a bit randomly, but I'll try without the requests one. I though mypy was supposed to handle cyclic references though, but maybe not at this level?

Reply:
Can you try this test:

```sh
pytest tests/typecheck/test_shortcuts.yml::get_user_model_returns_proper_class
```

It's one of the ones triggering the cyclic dependencies problem here. If fails even if I remove all references to `auth` in `requests.pyi`

Reply:
Hm, I don't hit any errors when trying it

Reply:
Alright, so my issue is that I'm checking if `django.contrib.auth` is installed, and if it's not I don't add the dependencies. Guess I'll have to properly figure out what do when that isn't installed. I believe it'll raise an error at runtime 🤔